src/goto-programs/goto_cfg.cpp

```diff
@@ -165,3 +194,646 @@ void goto_cfg::dump_graph() const
+    for (const auto &item : intersection)
+      result.insert(item);
+
+    // Fix-point?
+    if (dt[dominator_node] != result)
+    {
+      dt[dominator_node] = result;
+      for (const auto &suc : dominator_node->successors)
+        worklist.push_back(suc);
+    }
+  }
+
+  _dominators = dt;
+}
+
+void Dominator::dump_dominators() const
+{
+  for (const auto &[node, edges] : _dominators)
+  {
+    log_status("CFGNode");
+    node->begin->dump();
+
+    log_status("Dominated");
+    for (const auto &edge : edges)
+      edge->begin->dump();
+  }
+}
+
+CFGNode Dominator::idom(const CFGNode &n) const
+{
+  std::vector<CFGNode> sdoms;
+  const auto &n_dominators = dom(n);
+  std::copy_if(
+    n_dominators.begin(),
+    n_dominators.end(),
+    std::back_inserter(sdoms),
+    [&n, this](const auto &item) { return sdom(item, n); });
+
+  for (const CFGNode &n0 : sdoms)
+  {
+    bool valid_result = true;
+    for (const CFGNode &n1 : sdoms)
+    {
+      if (sdom(n0, n1))
+      {
+        valid_result = false;
+        break;
+      }
+    }
+    if (valid_result)
+      return n0;
+  }
+  log_error("[cfg] Unable to compute `idom`");
+  abort();
+}
+
+void Dominator::dump_idoms() const
+{
+  DomTree dt(*this);
+
+  log_status("Root: {}", dt.root->uuid);
+  for (const auto &[key, value] : dt.edges)
+  {
+    log_status("CFGNode: {}", key->uuid);
+    log_status("Edges");
+    for (const auto &n : value)
+      log_status("\t{}", n->uuid);
+  }
+}
+
+std::unordered_set<CFGNode> Dominator::dom_frontier(const CFGNode &n) const
+{
+  assert(dj);
+  std::unordered_set<CFGNode> result;
+  const auto levels = dj->tree.get_levels();
+  for (const CFGNode &y : dj->tree.get_subtree(n))
+  {
+    for (const CFGNode &z : dj->_jedges[y])
+      if (levels.at(z) <= levels.at(n))
+        result.insert(z);
+  }
+  return result;
+}
+
+std::unordered_set<CFGNode>
+Dominator::dom_frontier(const std::unordered_set<CFGNode> &nodes) const
+{
+  assert(nodes.size() > 0);
+
+  std::unordered_set<CFGNode> result;
+
+  for (const CFGNode &n : nodes)
+    for (const CFGNode &df : dom_frontier(n))
+      result.insert(df);
+
+  return result;
+}
+
+std::unordered_set<CFGNode>
+Dominator::iterated_dom_frontier(const std::unordered_set<CFGNode> &nodes) const
+{
+  assert(dj);
+  assert(nodes.size() > 0);
+  // TODO: There is a linear solution by SREEDHAR using PiggyBanks
+  std::unordered_set<CFGNode> result = dom_frontier(nodes);
+  std::unordered_set<CFGNode> result2 = dom_frontier(result);
+
+  while (1)
+  {
+    result = result2;
+    result2 = dom_frontier(result);
+    assert(result2.size() >= result.size());
+
+    if (result.size() != result2.size())
+      continue;
+
+    for (const auto &n : result)
+      if (!result2.count(n))
+        continue;
+
+    break;
+  }
+
+  return result;
+}
+
+Dominator::DJGraph::DJGraph(const CFGNode &cfg, const Dominator &dom)
+  : tree(dom)
+{
+  // A DJ-Graph is a graph composed by D-Edges and J-Edges
```

Comment:
Add a reference to a paper?

Reply:
The reference is at the declaration of the class